modules/linear_boltzmann_solvers/discrete_ordinates_solver/sweep/communicators/aah_async_comm.cc

```diff
@@ -70,192 +62,157 @@ AAH_ASynchronousCommunicator::Reset()
+  auto message_count_and_size = [this](const auto num_unknowns)
   {
-    size_t num_unknowns = fluds.GetPrelocIFaceDOFCount(prelocI) * num_groups_ * num_angles_;
     size_t message_count = num_angles_;
     if (num_unknowns * 8 > max_mpi_message_size_)
       message_count = ((num_unknowns * 8) + (max_mpi_message_size_ - 1)) / max_mpi_message_size_;
     size_t message_size = (num_unknowns + (message_count - 1)) / message_count;
+    return std::make_pair(message_count, message_size);
+  };
 
-    prelocI_message_count_[prelocI] = message_count;
-    prelocI_message_size_[prelocI].reserve(message_count);
-    prelocI_message_blockpos_[prelocI].reserve(message_count);
+  // Predecessor locations
+  const size_t num_dependencies = spds.GetLocationDependencies().size();
+  preloc_msg_data_.resize(num_dependencies);
+  preloc_msg_received_.resize(num_dependencies);
+
+  for (int i = 0; i < num_dependencies; ++i)
+  {
+    size_t num_unknowns = fluds.GetPrelocIFaceDOFCount(i) * num_groups_ * num_angles_;
+    auto [message_count, message_size] = message_count_and_size(num_unknowns);
+    const int locJ = spds.GetLocationDependencies()[i];
+    const auto source = comm_set_.MapIonJ(locJ, opensn::mpi_comm.rank());
 
     size_t pre_block_pos = 0;
+    preloc_msg_data_[i].reserve(message_count);
     for (int m = 0; m < message_count - 1; ++m)
     {
-      prelocI_message_size_[prelocI].push_back(message_size);
-      prelocI_message_blockpos_[prelocI].push_back(pre_block_pos);
+      preloc_msg_data_[i].emplace_back(std::make_tuple(source, message_size, pre_block_pos));
       num_unknowns -= message_size;
       pre_block_pos += message_size;
     }
     if (num_unknowns > 0)
-    {
-      prelocI_message_blockpos_[prelocI].push_back(pre_block_pos);
-      prelocI_message_size_[prelocI].push_back(num_unknowns);
-    }
+      preloc_msg_data_[i].emplace_back(std::make_tuple(source, num_unknowns, pre_block_pos));
+    else
+      --message_count;
 
-    prelocI_message_received_.emplace_back(message_count, false);
+    preloc_msg_received_[i].resize(message_count, false);
+    max_num_messages_ = std::max(message_count, max_num_messages_);
   }
 
-  // Delayed Predecessor locations
-  size_t num_delayed_dependencies = spds.GetDelayedLocationDependencies().size();
+  // Delayed predecessor locations
+  const size_t num_delayed_dependencies = spds.GetDelayedLocationDependencies().size();
+  delayed_preloc_msg_data_.resize(num_delayed_dependencies);
+  delayed_preloc_msg_received_.resize(num_delayed_dependencies);
 
-  delayed_prelocI_message_count_.resize(num_delayed_dependencies, 0);
-  delayed_prelocI_message_size_.resize(num_delayed_dependencies);
-  delayed_prelocI_message_blockpos_.resize(num_delayed_dependencies);
-  delayed_prelocI_message_received_.clear();
-
-  for (int prelocI = 0; prelocI < num_delayed_dependencies; ++prelocI)
+  for (int i = 0; i < num_delayed_dependencies; ++i)
```

Comment:
`i` should be `size_t` as well to match the `num_delayed_dependencies` type.

Reply:
Fixed.